File: frontend/lib/planning-revision.ts
import type { EpisodeRoadmapItem, ScriptProject, StoryTreeQualityAudit } from "./types";
import { CURRENT_STORY_REVIEW_CONTRACT_VERSION, type StoryPlanNode } from "./story-planning-client";

export function planningRevisionEpoch(project: Pick<ScriptProject, "planningRevisionEpoch">): number {
  return project.planningRevisionEpoch ?? 0;
}

export function isPlanningRevisionActive(project: Pick<ScriptProject, "planningRevision">): boolean {
  return project.planningRevision?.status === "active";
}

/** Include saved raw bodies and edited bodies; an unconfirmed body is still written. */
export function nextUnwrittenPlanningEpisode(project: Pick<ScriptProject, "episodes">): number {
  return Math.max(0, ...project.episodes.map((episode) => episode.episodeNumber)) + 1;
}

export function planningRevisionNodeLocked(project: ScriptProject, node: StoryPlanNode): boolean {
  const revision = project.planningRevision;
  return revision?.status === "active" && (node.planned_start_episode == null
    || node.planned_end_episode == null || node.planned_start_episode < revision.startEpisode);
}

export function planningRevisionEpisodeLocked(project: ScriptProject, episodeNumber: number): boolean {
  const revision = project.planningRevision;
  return revision?.status === "active" && episodeNumber < revision.startEpisode;
}

export function startPlanningRevision(project: ScriptProject, startEpisode = nextUnwrittenPlanningEpisode(project)): ScriptProject {
  if (isPlanningRevisionActive(project)) throw new Error("后续规划修订已经开始，请继续当前修订。");
  if (project.planningSession?.phase !== "script" || project.planningSession.status !== "approved") {
    throw new Error("请先确认当前规划，再开启后续修订。");
  }
  if (!Number.isInteger(startEpisode) || startEpisode < nextUnwrittenPlanningEpisode(project)
    || startEpisode > project.generationSettings.episodeCount) {
    throw new Error("只能修订尚无正文的后续集数。");
  }
  if (project.activeGenerationTask?.status === "running") {
    throw new Error("请等待当前生成任务停止并保存，再修订后续规划。");
  }
  const now = new Date().toISOString();
  return {
    ...project,
    planningRevisionEpoch: planningRevisionEpoch(project) + 1,
    planningRevision: {
      revisionId: `planning-revision.${crypto.randomUUID()}`,
      status: "active", startEpisode,
      sourceWorkspaceRevision: project.serverSync?.workspaceRevision ?? 0,
      startedAt: now,
      originalRoadmaps: structuredClone((project.episodeRoadmaps ?? []).filter((item) => item.episode_number >= startEpisode)),
      ...(project.storyTreeQualityAudit ? { originalAudit: structuredClone(project.storyTreeQualityAudit) } : {}),
      ...(project.planningSession ? { originalPlanningSession: structuredClone(project.planningSession) } : {}),
      invalidatedJobIds: project.activeGenerationTask ? [project.activeGenerationTask.jobId] : [],
    },
    planningRevisionHistory: [
      ...(project.planningRevisionHistory ?? []),
      ...(project.planningRevision ? [structuredClone(project.planningRevision)] : []),
    ],
    episodeRoadmaps: (project.episodeRoadmaps ?? []).map((item) => item.episode_number < startEpisode
      ? item : { ...item, status: "draft" }),
    episodePlansReadyThrough: Math.min(project.episodePlansReadyThrough ?? 0, startEpisode - 1) || undefined,
    storyTreeQualityAudit: undefined,
    activeGenerationTask: undefined,
    deliveryConfirmation: undefined,
    updatedAt: now,
  };
}

/** Recover the visible draft if a node save succeeded before its workspace checkpoint. */
export function revisionRoadmapsForNode(project: ScriptProject, node: StoryPlanNode): EpisodeRoadmapItem[] {
  const candidates = (project.episodeRoadmaps ?? []).filter((item) => item.source_node_id === node.node_id
    && item.story_bible_version === node.story_bible_version
    && (isPlanningRevisionActive(project) || item.source_node_version === node.version));
  const byEpisode = new Map<number, EpisodeRoadmapItem>();
  for (const item of candidates) {
    const previous = byEpisode.get(item.episode_number);
    if (!previous || item.source_node_version === node.version
      || (previous.source_node_version !== node.version && item.source_node_version > previous.source_node_version)) {
      byEpisode.set(item.episode_number, item);
    }
  }
  return [...byEpisode.values()].sort((left, right) => left.episode_number - right.episode_number);
}

/** A source identity match alone cannot make stale episode events approvable. */
export function planningRevisionSourceIssues(item: EpisodeRoadmapItem, node: StoryPlanNode): string[] {
  if (item.source_node_id !== node.node_id || item.source_node_version !== node.version
    || item.story_bible_version !== node.story_bible_version) return ["分集的上层来源已变化，请先复核来源。"];
  const event = node.episode_developments?.find((entry) => entry.episode_number === item.episode_number);
  if (!event) return ["上层节点缺少本集事件，请先补齐上层规划。"];
  const fields = ["entry_state", "exit_state", "source_turning_points", "source_unit_story_beats"] as const;
  const mismatches = fields.filter((field) => JSON.stringify(item[field]) !== JSON.stringify(event[field]));
  return mismatches.length ? ["本集的起止状态或来源事件与最新上层规划不一致，请先修订本集，再批准。"] : [];
}

/** During an explicit revision, downstream prose survives even a boundary edit. */
export function replaceRevisionRoadmap(project: ScriptProject, replacement: EpisodeRoadmapItem): EpisodeRoadmapItem[] {
  if (!isPlanningRevisionActive(project) || planningRevisionEpisodeLocked(project, replacement.episode_number)) {
    throw new Error("该集不在本次后续规划修订范围内。");
  }
  let replaced = false;
  const result = (project.episodeRoadmaps ?? []).map((item) => {
    if (item.episode_number === replacement.episode_number && item.source_node_id === replacement.source_node_id
      && item.source_node_version === replacement.source_node_version && item.story_bible_version === replacement.story_bible_version) {
      replaced = true;
      return { ...replacement, status: "draft" as const };
    }
    return item.episode_number > replacement.episode_number ? { ...item, status: "draft" as const } : item;
  });
  if (!replaced) throw new Error("本集来源已经变化，请重新加载当前规划后再修改。");
  return result;
}

/** Preserve every text field while making a changed source explicitly reviewable. */
export function retainRevisionRoadmaps(
  project: ScriptProject, previousVersions: Map<string, number>, nextVersions: Map<string, number>,
): EpisodeRoadmapItem[] {
  const changedFrom = Math.min(...(project.episodeRoadmaps ?? []).filter((item) => (
    previousVersions.get(item.source_node_id) === item.source_node_version
    && nextVersions.get(item.source_node_id) != null
    && nextVersions.get(item.source_node_id) !== item.source_node_version
  )).map((item) => item.episode_number));
  return (project.episodeRoadmaps ?? []).map((item) => {
    if (planningRevisionEpisodeLocked(project, item.episode_number)) return item;
    const previous = previousVersions.get(item.source_node_id);
    const next = nextVersions.get(item.source_node_id);
    if (previous !== item.source_node_version || next == null || next === previous) {
      return item.episode_number >= changedFrom ? { ...item, status: "draft" as const } : item;
    }
    return {
      ...item, source_node_version: next, status: "draft" as const,
      source_revision_review: { previous_version: previous, current_version: next },
    };
  });
}

export function completePlanningRevision(
  project: ScriptProject, nodes: StoryPlanNode[], audit: StoryTreeQualityAudit,
): ScriptProject {
  const revision = project.planningRevision;
  if (revision?.status !== "active") throw new Error("当前没有进行中的后续规划修订。");
  const scope = audit.future_revision_review;
  if (!scope || scope.revision_id !== revision.revisionId
    || scope.planning_revision_epoch !== planningRevisionEpoch(project)
    || scope.start_episode !== revision.startEpisode
    || scope.end_episode !== project.generationSettings.episodeCount
    || scope.status !== "pass" || scope.boundary_status !== "pass"
    || audit.findings.some(item => item.end_episode >= revision.startEpisode)
    || audit.review_contract_version !== CURRENT_STORY_REVIEW_CONTRACT_VERSION
    || !Number.isFinite(Date.parse(audit.created_at))
    || Date.parse(audit.created_at) < Date.parse(revision.startedAt)) {
    throw new Error("请完成本次修订后的完整剧情审校，确认未来范围及已保存正文衔接通过，再结束修订。");
  }
  for (let number = revision.startEpisode; number <= project.generationSettings.episodeCount; number += 1) {
    const matches = (project.episodeRoadmaps ?? []).filter((item) => item.episode_number === number
      && nodes.some((node) => node.node_id === item.source_node_id && node.version === item.source_node_version));
    const node = nodes.find((entry) => entry.node_id === matches[0]?.source_node_id
      && entry.version === matches[0]?.source_node_version);
    if (matches.length !== 1 || matches[0].status !== "approved" || matches[0].source_revision_review
      || !node || planningRevisionSourceIssues(matches[0], node).length) {
      throw new Error(`请先复核并逐集批准第${number}集规划。`);
    }
  }
  const now = new Date().toISOString();
  return {
    ...project, storyTreeQualityAudit: audit,
    planningRevision: { ...revision, status: "completed", completedAt: now },
    updatedAt: now,
  };
}

/** Do not expose an unlocked local state until the server accepts the transition. */
export async function persistPlanningRevisionTransition(
  project: ScriptProject,
  next: ScriptProject,
  save: (source: ScriptProject, candidate: ScriptProject) => Promise<ScriptProject>,
  apply: (saved: ScriptProject, source: ScriptProject) => Promise<unknown>,
): Promise<ScriptProject> {
  const saved = await save(project, next);
  if (await apply(saved, project) === false) throw new Error("服务器已保存本次修订，但本地未能更新，请重新加载项目。");
  return saved;
}

File: backend/app/modules/script_engine/planning_revision.py
"""Durable, future-only planning revisions; callers hold the project row lock."""

from __future__ import annotations

import json
from collections.abc import Mapping
from datetime import datetime, timezone

from app.modules.script_engine.long_story_repository import LongStoryPersistenceConflictError
from app.modules.script_engine.produced_plan_amendment import pending_episode_numbers
from app.modules.script_engine.planning_review_cache import CURRENT_REVIEW_CONTRACT_VERSION, quality_episode_projection, quality_node_signature
from app.modules.script_engine.future_revision_review import scope_matches_workspace


def revision_epoch(workspace: Mapping[str, object]) -> int:
    value = workspace.get("planningRevisionEpoch", 0)
    if type(value) is not int or value < 0:
        raise LongStoryPersistenceConflictError("Invalid planning revision epoch.")
    return value


def active_revision(workspace: Mapping[str, object]) -> dict | None:
    marker = workspace.get("planningRevision")
    return marker if isinstance(marker, dict) and marker.get("status") == "active" else None


def require_request_epoch(workspace: Mapping[str, object], requested: int, *, body: bool = False,
                          episode_number: int | None = None) -> None:
    if requested != revision_epoch(workspace):
        raise LongStoryPersistenceConflictError("Planning revision epoch changed; reload before continuing.")
    pending = pending_episode_numbers(workspace)
    if body and pending and (episode_number is None or episode_number >= pending[0]):
        raise LongStoryPersistenceConflictError("Produced planning changed; resolve affected screenplays before generating or recovering later episodes.")
    active = active_revision(workspace)
    if active and body:
        raise LongStoryPersistenceConflictError("Planning revision is active; screenplay generation and recovery are blocked.")
    if active and episode_number is not None and episode_number < active["startEpisode"]:
        raise LongStoryPersistenceConflictError("Planning revision cannot change episodes before its frozen boundary.")


def _fail(message: str) -> None:
    raise LongStoryPersistenceConflictError(f"Planning revision: {message}")


def _timestamp(value: object) -> datetime:
    try:
        result = datetime.fromisoformat(str(value).replace("Z", "+00:00"))
        if result.tzinfo is None:
            raise ValueError("Timezone required")
        return result
    except (ValueError, TypeError):
        _fail("a valid timezone-aware timestamp is required.")


def _roadmaps(workspace: Mapping[str, object]) -> dict[int, dict]:
    rows = workspace.get("episodeRoadmaps", [])
    if not isinstance(rows, list) or any(not isinstance(row, dict) or type(row.get("episode_number")) is not int for row in rows):
        _fail("roadmaps must retain their episode identities.")
    result = {row["episode_number"]: row for row in rows}
    if len(result) != len(rows):
        _fail("duplicate roadmap episode identities are not allowed.")
    return result


def validate_workspace_transition(previous: Mapping[str, object], candidate: Mapping[str, object], *,
                                  source_revision: int, planned_episode_count: int,
                                  saved_through_episode: int, has_running_work: bool, produced_plan_transition: str | None = None) -> str | None:
    """Return start/complete only after validating immutable source and boundaries."""
    old_epoch, new_epoch = revision_epoch(previous), revision_epoch(candidate)
    old, new = previous.get("planningRevision"), candidate.get("planningRevision")
    history, next_history = previous.get("planningRevisionHistory", []), candidate.get("planningRevisionHistory", [])
    if not old and not new and not old_epoch and not new_epoch and history == next_history == []:
        return None
    if produced_plan_transition == "amend":
        if new_epoch != old_epoch + 1 or old != new or history != next_history:
            _fail("execution amendments must retain the separate future revision and history.")
        return "amend"
    if not isinstance(new, dict) or not isinstance(next_history, list):
        _fail("the current revision and its history cannot be removed.")
    starting = new_epoch == old_epoch + 1
    completing = isinstance(old, dict) and old.get("status") == "active" and new.get("status") == "completed"
    if starting:
        if active_revision(previous):
            _fail("finish the current revision before starting another.")
        if has_running_work:
            _fail("running generation or review work must finish before reopening planning.")
        session = previous.get("planningSession") or {}
        if not isinstance(session, dict) or session.get("phase") != "script" or session.get("status") != "approved":
            _fail("only confirmed planning in the script phase can be reopened.")
        expected_history = history + ([old] if isinstance(old, dict) else [])
        if next_history != expected_history:
            _fail("revision history must preserve the prior completed record exactly once.")
        if new.get("status") != "active" or not isinstance(new.get("revisionId"), str) or not new["revisionId"]:
            _fail("a new active revision must identify the next epoch.")
        if new.get("revisionId") in [item.get("revisionId") for item in expected_history if isinstance(item, dict)]:
            _fail("revision identities cannot be reused.")
        if new.get("sourceWorkspaceRevision") != source_revision:
            _fail("source workspace revision is stale; reload before reopening.")
        boundary = new.get("startEpisode")
        if type(boundary) is not int or not saved_through_episode < boundary <= planned_episode_count:
            _fail("start episode must be after all saved screenplays and inside the project.")
        _timestamp(new.get("startedAt"))
        if "completedAt" in new:
            _fail("an active revision cannot already be completed.")
        expected_tail = [row for row in previous.get("episodeRoadmaps", []) if row["episode_number"] >= boundary]
        if new.get("originalRoadmaps") != expected_tail:
            _fail("original future roadmaps must match the saved source exactly.")
        for snapshot, field in (("originalAudit", "storyTreeQualityAudit"), ("originalPlanningSession", "planningSession")):
            if new.get(snapshot) != previous.get(field):
                _fail("original audit and planning session must match the saved source exactly.")
        expected_jobs = [previous["activeGenerationTask"]["jobId"]] if previous.get("activeGenerationTask") else []
        if new.get("invalidatedJobIds", []) != expected_jobs or candidate.get("activeGenerationTask") is not None:
            _fail("reopening must invalidate and retain the identity of the previous generation task.")
        if candidate.get("storyTreeQualityAudit") is not None:
            _fail("a reopened revision requires a new full audit.")
    else:
        if new_epoch != old_epoch or next_history != history or not isinstance(old, dict):
            _fail("epoch and historical snapshots cannot be altered by an ordinary save.")
        mutable = {"status", "completedAt"} if completing else set()
        if {k: v for k, v in old.items() if k not in mutable} != {k: v for k, v in new.items() if k not in mutable}:
            _fail("the original revision record is immutable.")
        if not completing and old != new:
            _fail("invalid revision status transition.")
        boundary = new.get("startEpisode")
        if completing:
            if has_running_work:
                _fail("running planning work must finish before completing the revision.")
            if _timestamp(new.get("completedAt")) < _timestamp(new.get("startedAt")):
                _fail("completion cannot predate the revision.")
    # A completed record protects its historical metadata, not unrelated future
    # authoring workflows. A new planning revision has its own explicit opening.
    if not starting and not completing and not active_revision(previous):
        return None
    if candidate.get("episodes", []) != previous.get("episodes", []):
        # Once closed, normal screenplay saves can resume with the current epoch.
        if starting or completing or active_revision(previous):
            _fail("existing screenplays must be preserved and new screenplay saves are blocked while planning is open.")
    before, after = _roadmaps(previous), _roadmaps(candidate)
    if set(before) != set(after):
        _fail("reopening planning cannot delete or replace the future roadmap range.")
    for number, row in after.items():
        if number < boundary and row != before[number]:
            source = next((e.get("sourceAmendment") for e in previous.get("episodes", []) if e.get("episodeNumber") == number), None)
            status_only_approval = (isinstance(source, dict) and source.get("status") == "revision_required"
                                    and before[number].get("status") == "draft" and row == {**before[number], "status": "approved"})
            if not status_only_approval:
                _fail("roadmaps before the revision boundary are frozen.")
        if starting and number >= boundary and row != {**before[number], "status": "draft"}:
            _fail("opening must preserve each future roadmap as an unapproved draft.")
        if row.get("status") == "approved" and row.get("source_revision_review"):
            _fail("a changed source node must be explicitly reviewed before approving its roadmap.")
    editable = {"planningRevisionEpoch", "planningRevision", "planningRevisionHistory", "episodeRoadmaps",
                "episodePlansReadyThrough", "storyTreeQualityAudit", "activeGenerationTask", "deliveryConfirmation",
                "updatedAt", "serverSync", "deliveryContentRevision", "activeEpisodeNumber"}
    for field in set(previous) | set(candidate):
        if field not in editable and candidate.get(field) != previous.get(field):
            _fail(f"only future planning may change during this revision ({field}).")
    if candidate.get("activeGenerationTask") is not None:
        _fail("generation tasks cannot be restored while the revision is active.")
    session = candidate.get("planningSession") or {}
    if new.get("status") == "active" and (not isinstance(session, dict) or session.get("phase") != "script" or session.get("status") != "approved"):
        _fail("the confirmed planning session is preserved while the separate revision is active.")
    if completing:
        if session.get("phase") != "script" or session.get("status") != "approved":
            _fail("completed revisions must return to approved script planning.")
        require_fresh_complete_audit(candidate, planned_episode_count=planned_episode_count)
    return "start" if starting else "complete" if completing else None


def require_fresh_complete_audit(workspace: Mapping[str, object], *, planned_episode_count: int) -> None:
    audit, revision = workspace.get("storyTreeQualityAudit"), workspace.get("planningRevision")
    plans = _roadmaps(workspace)
    if set(plans) != set(range(1, planned_episode_count + 1)) or any(row.get("status") != "approved" or row.get("source_revision_review") for row in plans.values()):
        _fail("all episodes must retain current, individually approved roadmaps.")
    if not isinstance(audit, dict) or not isinstance(revision, dict) or audit.get("review_contract_version") != CURRENT_REVIEW_CONTRACT_VERSION:
        _fail("a fresh full current-contract audit is required.")
    scope = audit.get("future_revision_review")
    if not scope_matches_workspace(scope, workspace) or scope.get("status") != "pass" or scope.get("boundary_status") != "pass":
        _fail("a fresh server-bound future-range and saved-boundary review PASS is required.")
    if any(item.get("end_episode", 2_001) >= revision["startEpisode"] for item in audit.get("findings", [])):
        _fail("unresolved future-range findings still block completion.")
    if _timestamp(audit.get("created_at")) < _timestamp(revision["startedAt"]):
        _fail("the audit predates this planning revision.")
    if audit.get("story_bible_version") != workspace.get("storyBibleVersion"):
        _fail("the audit Story Bible is stale.")
    refs = {(item.get("source_node_id"), item.get("source_node_version")) for item in plans.values()}
    raw_refs = audit.get("node_refs")
    if not isinstance(raw_refs, list) or any(not isinstance(item, dict) for item in raw_refs):
        _fail("full audit node evidence is missing.")
    audit_refs = {(item.get("node_id"), item.get("node_version")) for item in raw_refs}
    if audit_refs != refs or len(raw_refs) != len(refs) or audit.get("audited_node_count") != len(refs) or audit.get("semantic_sample_count") != len(refs) or audit.get("node_signature") != quality_node_signature(refs):
        _fail("the audit must cover the complete current node frontier.")
    try:
        reviewed = json.loads(audit.get("reviewed_episode_plans") or "null")
        actual = [quality_episode_projection(plans[number]).model_dump(mode="json") for number in sorted(plans)]
    except (TypeError, ValueError, KeyError):
        _fail("full reviewed episode evidence is invalid.")
    if reviewed != actual:
        _fail("the approved roadmaps differ from the audit's exact episode evidence.")
